cli, cli/sql: misc improvements

knz opened the pull request:
Fixes  #19245.
Informs #19445.

- This ensures that a 1.1+ client can connect to a 1.0 server.
- prints a warning when the server version is older, as requested in #19245.
- also renames `--unsafe-updates` (with default false) to `--safe-updates` (with default true)
- fixes a bug introduced two days ago

Will cherry-pick at least the 1st and last commit into 1.1. 

Should I also cherry-pick the other 3? In particular the 3rd which renames the flag?

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20935)
<!-- Reviewable:end -->

knz committed:
cli/sql: do not fail if sql_safe_updates cannot be set

Prior to this patch, an interactive SQL shell (`cockroach sql`) would
error out if connecting to a server which doesn't support the session
variable `sql_safe_updates` -- e.g. CockroachDB 1.0 or PostgreSQL
itself

This patch ensures that a failure is only reported as a warning.

Release note (cli change): `cockroach sql` does not fail any more when
the server does not support the `sql_safe_updates` session variable.

knz committed:
cli: de-emphasize the seriousness of errors in ancillary SQL client services

The code that implements the SQL connection object utilized by various
CLI utilities (including, but not limited to, `cockroach sql`) provide
ancillary services for the user's convenience, like checking whether
the client and the server version are the same, etc.

When these checks fail, the failure should not indicate malfunction --
they can simply indicate that the utility was run against a previous
CockroachDB version that does not yet support some service, or that
`cockroach sql` is being run against a PostgreSQL server.

Prior to this patch, the errors due to ancillary failures were
reported with serious-sounding messages, often starting with the word
"unable" or "error".

This patch de-emphasizes them by prefixing them with the string
"warning:".

Release note (cli change): some warning messages are now more clearly
indicated with the "warning:" prefix.

knz committed:
cli/sql: rename `--unsafe-updates` to `--safe-updates`

Prior to this patch, the session variable was called
`sql_safe_updates` (aligned with MySQL) but the command-line flag
`--unsafe-updates` (defaulting to false). This inversion of semantics
in the flag vs. the session variable was difficult to comprehend and
makes the feature difficult to document.

This patch alleviates this difficulty by flipping the meaning of the
flag, renaming it to `--safe-updates` with a default of true.

Release note (cli change): the command-line flag `--unsafe-updates` is
renamed to `--safe-updates`. The default behavior is unchanged: the
previous flag defaulted to false, the new one to true.

knz committed:
cli/sql: ensure that --safe-updates works even in non-interactive sessions

The intended behavior of `--safe-updates`, when specified explicitly,
is to disallow "dangerous statements". This feature should work
whenever the flag is explicitly specified, regardless of whether the
session is interactive.

Prior to this patch, the code was erroneously ignoring the
`--safe-updates` flag in non-interactive session, causing it to become
ineffective when e.g. combined with `-e`. This patch fixes the
behavior.

Release note (cli change, bug fix): `cockroach sql --safe-updates` can
now be used effectively in non-interactive sessions.

Note: this fixes a regression introduced in
475e0a30726b5a72940d94e9b419e5ef08a79146. The flag was functioning
properly before that.

knz committed:
cli: add a check that client version <= server version

This patch introduces two features:

- it ensures that a v1.1+ `cockroach` client connecting to a v1.0
  server properly recognizes that the server is running v1.0 (prior to
  this patch, the version strings from a v1.0 server were not
  recognized).

- it causes a warning message to be printed if the client is newer
  than the server.

For example:

```
kena@kenax ~/cockroach % ./cockroach sql --insecure

warning: server version older than client! proceed with caution; some features may not be available.

root@:26257/>
```

Release note (cli change): client commands that use SQL (including
`cockroach sql`, `cockroach node ls`, etc) now print a warning if the
server version is older than the client.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

I would cherry-pick the first two and the last, but not the renaming of the flag (and if we don't cherry-pick r3, then I don't think r4 is worth cherry-picking on its own). 

---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20935#-:-L0pbHE260zDS8rKyj3j:b5e5oxs)*
<!-- Sent from Reviewable.io -->

knz commented:
TFYR, Done.

knz committed:
cli/interactive_tests: fix typo in test controller

Release note: None
